helix-core/src/history.rs

```diff
@@ -22,72 +58,195 @@ impl Default for History {
-            // We're at the root of undo, nothing to do.
             return None;
         }
 
-        let current_revision = &self.revisions[self.cursor];
+        let current_revision = &self.revisions[self.current];
+        self.current = current_revision.parent;
+        Some(&current_revision.inversion)
+    }
 
-        self.cursor = current_revision.parent;
+    pub fn redo(&mut self) -> Option<&Transaction> {
+        let current_revision = &self.revisions[self.current];
+        let last_child = current_revision.last_child;
+        if last_child == 0 {
+            return None;
+        }
+        self.current = last_child;
 
-        Some(&current_revision.revert)
+        let last_child_revision = &self.revisions[last_child];
+        Some(&self.revisions[last_child].transaction)
     }
 
-    pub fn redo(&mut self) -> Option<&Transaction> {
-        let current_revision = &self.revisions[self.cursor];
+    fn lowest_common_ancestor(&self, mut a: usize, mut b: usize) -> usize {
+        use std::collections::HashSet;
+        let mut a_path_set = HashSet::new();
+        let mut b_path_set = HashSet::new();
+        loop {
+            a_path_set.insert(a);
+            b_path_set.insert(b);
+            if a_path_set.contains(&b) {
+                return b;
+            }
+            if b_path_set.contains(&a) {
+                return a;
+            }
+            a = self.revisions[a].parent; // Relies on the parent of 0 being 0.
+            b = self.revisions[b].parent; // Same as above.
+        }
+    }
 
-        // for now, simply pick the latest child (linear undo / redo)
-        if let Some((index, transaction)) = current_revision.children.last() {
-            self.cursor = *index;
+    // List of nodes on the way from `n` to 'a`. Doesn`t include `a`.
+    // Includes `n` unless `a == n`. `a` must be an ancestor of `n`.
+    fn path_up(&self, mut n: usize, a: usize) -> Vec<usize> {
+        let mut path = Vec::new();
+        while n != a {
+            path.push(n);
+            n = self.revisions[n].parent;
+        }
+        path
+    }
+
+    fn jump_to(&mut self, to: usize) -> Vec<Transaction> {
+        let lca = self.lowest_common_ancestor(self.current, to);
+        let up = self.path_up(self.current, lca);
+        let down = self.path_up(to, lca);
+        self.current = to;
+        let up_txns = up.iter().map(|&n| self.revisions[n].inversion.clone());
+        let down_txns = down
+            .iter()
+            .rev()
+            .map(|&n| self.revisions[n].transaction.clone());
+        up_txns.chain(down_txns).collect()
+    }
+
+    fn jump_backward(&mut self, delta: usize) -> Vec<Transaction> {
+        self.jump_to(self.current.saturating_sub(delta))
+    }
+
+    fn jump_forward(&mut self, delta: usize) -> Vec<Transaction> {
+        use std::cmp::max;
+        self.jump_to(max(
+            self.revisions.len() - 1,
+            self.current.saturating_add(delta),
+        ))
```

Comment:
I think this should be `min` right? It panics when number is too large. Also, I did some cleaning here.
```suggestion
        self.jump_to(self.current.saturating_add(delta).min(self.revisions.len() - 1))
```
```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1000', helix-core/src/history.rs:137:17
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:69:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/index.rs:182:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/index.rs:15:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:2381:9
   6: helix_core::history::History::lowest_common_ancestor
             at ./helix-core/src/history.rs:137:17
   7: helix_core::history::History::jump_to
             at ./helix-core/src/history.rs:153:19
   8: helix_core::history::History::jump_forward
             at ./helix-core/src/history.rs:171:9
   9: helix_core::history::History::later
             at ./helix-core/src/history.rs:228:25
  10: helix_view::document::Document::later
             at ./helix-view/src/document.rs:405:20
  11: helix_term::commands::cmd::later
             at ./helix-term/src/commands.rs:1047:9
  12: helix_term::commands::command_mode::{{closure}}
             at ./helix-term/src/commands.rs:1178:17
  13: <alloc::boxed::Box<F,A> as core::ops::function::FnMut<Args>>::call_mut
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1553:9
  14: <helix_term::ui::prompt::Prompt as helix_term::compositor::Component>::handle_event
             at ./helix-term/src/ui/prompt.rs:275:21
  15: helix_term::compositor::Compositor::handle_event
             at ./helix-term/src/compositor.rs:112:19
  16: helix_term::application::Application::handle_terminal_events
             at ./helix-term/src/application.rs:144:32
  17: helix_term::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:108:21
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  19: helix_term::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:266:9
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  21: hx::main::{{closure}}
             at ./helix-term/src/main.rs:96:5
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  23: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:54
  24: tokio::coop::with_budget::{{closure}}
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:106:9
  25: std::thread::local::LocalKey<T>::try_with
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  26: std::thread::local::LocalKey<T>::with
             at /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  27: tokio::coop::with_budget
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:99:5
  28: tokio::coop::budget
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:76:5
  29: tokio::park::thread::CachedParkThread::block_on
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:31
  30: tokio::runtime::enter::Enter::block_on
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/enter.rs:151:13
  31: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/thread_pool/mod.rs:71:9
  32: tokio::runtime::Runtime::block_on
             at /home/ivan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/mod.rs:452:43
  33: hx::main
             at ./helix-term/src/main.rs:98:5
  34: core::ops::function::FnOnce::call_once
```

Reply:
Good catch!